src/rumor_state.rs

```diff
@@ -0,0 +1,177 @@
+                if round >= max_rounds {
+                    return RumorState::D;
+                }
+
+                // For any `peers_in_this_round` which aren't accounted for in `peer_ages`, add
+                // a age of `0` for them to indicate they're in state A (i.e. they didn't have
+                // the rumor).
+                for peer in peers_in_this_round {
+                    if let Entry::Vacant(entry) = peer_ages.entry(*peer) {
+                        let _ = entry.insert(0);
+                    }
+                }
+
+                // Apply the median rule, but if any peer's age >= `max_b_age` (i.e. that peer
+                // is in state C), transition to state C.
+                let mut less = 0;
+                let mut greater_or_equal = 0;
+                for peer_age in peer_ages.values() {
+                    if *peer_age < our_age {
+                        less += 1;
+                    } else if *peer_age >= max_b_age {
+                        return RumorState::C {
+                            rounds_in_state_b: round,
+                            round: 0,
+                        };
+                    } else {
+                        greater_or_equal += 1;
+                    }
+                }
+                if greater_or_equal > less {
+                    our_age += 1;
+                }
+
+                // If our age has reached `max_b_age`, transition to state C, otherwise remain
+                // in state B.
+                if our_age >= max_b_age {
+                    return RumorState::C {
+                        rounds_in_state_b: round,
+                        round: 0,
+                    };
+                }
+                RumorState::B {
+                    round,
+                    our_age,
+                    peer_ages: BTreeMap::new(),
+                }
+            }
+            RumorState::C {
+                rounds_in_state_b,
+                mut round,
+            } => {
+                round += 1;
+                // If we've hit the maximum permitted number of rounds, transition to state D
+                if round + rounds_in_state_b >= max_rounds {
+                    return RumorState::D;
+                }
+
+                // If we've hit the maximum rounds for remaining in state C, transition to state D.
+                if round >= max_c_rounds {
+                    return RumorState::D;
+                }
+
+                // Otherwise remain in state C.
+                RumorState::C {
+                    rounds_in_state_b,
+                    round,
+                }
+            }
+            RumorState::D => RumorState::D,
+        }
+    }
+
+    /// We only need to push and pull this rumor if we're in states B or C, hence this returns
+    /// `None` if we're in state D.  State C is indicated by returning a value > `max_b_age`.
+    pub fn our_age(&self) -> Option<u8> {
```

Comment:
Not sure of the name `our_age` ? Perhaps a slight refactor here to only allow rumor age to be retrieved in a cleaner way (as only B really returns a true age). I need to look deeper at the code locally though. In any case a follow-up PR, if at all.